### ISSUE TITLE
Future-proofing size_t namespace issues

### ISSUE THREAD

**Issue by jngrad** [espressomd/espresso#4313]
**TL;DR:** using plain `size_t` or `std::size_t` without the correct header file (`<stddef.h>` resp. `<cstddef>`) leads to portability issues, especially with recent compilers like GCC 10+ or Clang 11+. Can be mitigated by using `std::size_t` instead of `size_t`.

**Problem statement**

The codebase often relies on indirect includes from standard library and STL header files to make the `std::size_t` type visible. The correct way to use this type is to include one of the standard library header files that declares it, e.g. `<cstddef>, <cstdio>, <cstdlib>, <cstring>, <ctime>, <cuchar>, <cwchar>`. Relying on indirect includes can cause builds to fail with STL versions and standard library versions that are not available in our CI, since these libraries don't guarantee which headers are indirectly included in their header files.

In addition, the [C-compatibility headers](https://en.cppreference.com/w/cpp/header#C_compatibility_headers) like `stddef.h` are deprecated since C++11, which means `std::size_t` is the only future-proof way of using the type. Standard library and STL header files may sometimes provide the `size_t` type (without a namespace). The same issue applies to `std::ptrdiff_t`.

Portability issues for `size_t` depend on the `libc++` version, compiler vendor, compiler version and C++ standard.

MWE:

```c++
#include <array>
void foo(size_t arg);
```
```
$> clang++-10 -std=c++14 -c mwe.cpp
$> clang++-11 -std=c++14 -c mwe.cpp
mwe.cpp:2:10: error: unknown type name 'size_t'; did you mean 'std::size_t'?
$> clang++-12 -std=c++14 -c mwe.cpp
mwe.cpp:2:10: error: unknown type name 'size_t'; did you mean 'std::size_t'?
$> clang++-12 -std=c++20 -c mwe.cpp
$> gcc-9.4  -std=c++14 -c mwe.cpp
$> gcc-10.1 -std=c++14 -c mwe.cpp
mwe.cpp:2:10: error: 'size_t' was not declared in this scope; did you mean 'std::size_t'?
$> gcc-10.1 -std=c++17 -c mwe.cpp
mwe.cpp:2:10: error: 'size_t' was not declared in this scope; did you mean 'std::size_t'?
$> gcc-10.1 -std=c++20 -c mwe.cpp
$> gcc-11.1 -std=c++20 -c mwe.cpp
mwe.cpp:2:10: error: 'size_t' was not declared in this scope; did you mean 'std::size_t'?
```

**Current status**

There are 400 occurrences of plain `size_t` in ESPResSo according to `grep -rF size_t src | grep -vF 'std::size_t' | grep -vP '[a-z]size_t' | grep -vF '.pxd:' | grep -P 'size_t(?![a-zA-Z0-9_])'`. Older ESPResSo releases are also affected. In ESPResSo dependencies, there are 7 occurrences in Stokesian Dynamics and 3760 in waLBerla.

I tried a few combinations of compilers and compiler versions (GCC 10, GCC 11, Clang 12) and C++ standards (C++14, C++17) and couldn't get ESPResSo to fail a build. This is probably because `Utils::Vector3d` is used in many places, causing the indirect inclusion of `<cstddef>` in most of the codebase. In fact, on the tested build environments, `<cstddef>` includes both `<stddef.h>` (typedef for the plain `size_t`) and `<c++config.h>` (typedef for the namespaced version). This is probably for backward compatibility with C++ code that contains C code. I couldn't find information on how long this would remain so. For what's worth, there are examples involving the `musl` library that lead to portability issues with `<stddef.h>` ([mail for Clang 8](https://lists.llvm.org/pipermail/llvm-dev/2019-August/134350.html), [ticket for GCC 4.9](https://bugs.gentoo.org/603626)).

**Proposed solution**

Replacing all occurrences of plain `size_t` with `std::size_t` in ESPResSo and its dependencies should shield us from future portability issues. This can be achieved with sed expressions and the auto-formatting script.

For old ESPResSo releases, here is a quick fix: add the following code in the header file `config.hpp`, which is included directly or indirectly in almost all source files. This will make both `std::size_t` and `size_t` visible everywhere.
```c++
#include <cstddef>
using std::size_t;
```

**Comment by jngrad** [comment]
Fixed in waLBerla by [!478](https://i10git.cs.fau.de/walberla/walberla/-/merge_requests/478).
